test_conformance/images/kernel_read_write/test_cl_ext_image_from_buffer.cpp

```diff
@@ -79,77 +79,69 @@ int image2d_from_buffer_positive(cl_device_id device, cl_context context,
+                                     &row_pitch_alignment_2d, nullptr);
+        test_error(err, "Error clGetDeviceInfo");
 
-            cl_uint base_address_alignment_2d = 0;
-            err =
-                clGetDeviceInfo(device, CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT,
-                                sizeof(base_address_alignment_2d),
-                                &base_address_alignment_2d, nullptr);
-            test_error(err, "Error clGetDeviceInfo");
+        cl_uint base_address_alignment_2d = 0;
+        err = clGetDeviceInfo(device, CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT,
+                              sizeof(base_address_alignment_2d),
+                              &base_address_alignment_2d, nullptr);
+        test_error(err, "Error clGetDeviceInfo");
 
-            for (auto format : formatList)
-            {
-                cl_image_desc image_desc = { 0 };
-                image_desc_init(&image_desc, imageType);
+        for (auto format : formatList)
+        {
+            cl_image_desc image_desc = { 0 };
+            image_desc_init(&image_desc, imageType);
 
-                cl_mem_flags flag = (flagType == CL_MEM_KERNEL_READ_AND_WRITE)
-                    ? CL_MEM_READ_WRITE
-                    : flagType;
+            cl_mem_flags flag = (flagType == CL_MEM_KERNEL_READ_AND_WRITE)
+                ? CL_MEM_READ_WRITE
+                : flagType;
 
-                size_t row_pitch_alignment = 0;
-                size_t base_address_alignment = 0;
+            size_t row_pitch_alignment = 0;
+            size_t base_address_alignment = 0;
 
-                int get_error = get_image_requirement_alignment(
-                    device, context, flag, &format, &image_desc,
-                    &row_pitch_alignment, nullptr, &base_address_alignment);
-                if (TEST_PASS != get_error)
-                {
-                    return get_error;
-                }
+            int get_error = get_image_requirement_alignment(
+                device, context, flag, &format, &image_desc,
+                &row_pitch_alignment, nullptr, &base_address_alignment);
+            if (TEST_PASS != get_error)
+            {
+                return get_error;
+            }
 
-                const size_t element_size =
-                    get_format_size(context, &format, imageType, flag);
+            const size_t element_size =
+                get_format_size(context, &format, imageType, flag);
 
-                /*  Alignements in pixels vs bytes */
-                if (base_address_alignment
-                    > base_address_alignment_2d * element_size)
-                {
-                    test_fail("Unexpected base_address_alignment\n");
-                }
+            /*  Alignments in pixels vs bytes */
+            if (base_address_alignment
+                < base_address_alignment_2d * element_size)
+            {
+                test_fail("Unexpected base_address_alignment\n");
+            }
```

Comment:
At least according to my understanding this check is inverted, and I think this should be failing if the queried base_address_alignment is **greater than** the generic base_address_aligment_2d.  This is why the image requirements info extension is valuable - instead of having a single, generic per-device query that must return the worst-case scenario an application can instead query the specific requirements for an image.

```suggestion
            if (base_address_alignment
                > base_address_alignment_2d * element_size)
            {
                test_fail("Unexpected base_address_alignment\n");
            }
```

Reply:
hmmm. It looks like I misunderstood the extension, thank you for clearing it up. But with this interpretation it looks like @bcalidas was correct in pointing out that the comparison is flawed to begin with. 
The issue is that base_address_alignment returned by the extension is in bytes and has to be a power of 2. The right-hand part of the comparison is a calculation yielding a value not round up to a power of 2.
So say for a device where base_address_alignment_2d=32 pixels
for this format  ( CL_RGB, CL_UNORM_INT8) for instance 
* The extension could return  base_address_alignment = round_up_pow_of_2(32*3) = 128 
* The generic device would return base_address_alignment_2d * element_size=32*3 = 96

Should I round up (base_address_alignment_2d * element_size) to the closest power of 2? or drop the comparison

Reply:
Maybe to level set what we are comparing:

`base_address_alignment_2d` is the result of the query for [`CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT`](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT), which is:

> This query specifies the minimum alignment in pixels of the host_ptr specified to clCreateBuffer or clCreateBufferWithProperties when a 2D image is created from a buffer which was created using CL_MEM_USE_HOST_PTR. The value returned must be a power of 2.

`base_address_alignment` is the result of the query for [`CL_IMAGE_REQUIREMENTS_BASE_ADDRESS_ALIGNMENT_EXT`](https://registry.khronos.org/OpenCL/extensions/ext/cl_ext_image_requirements_info.html#clGetImageRequirementsInfoEXT), which is:

> Returns the minimum alignment in bytes required for the data store backing an image created using the parameters passed to clGetImageRequirementsInfoEXT. image_format or image_desc are allowed to be NULL. When either or both is NULL the implementation will return an alignment that would be sufficient for all possible values of the missing argument.
The value returned is a power of two.

So, both queries have a power of two requirement, but one is in units of pixels and another is in units of bytes.  I could see how this might be a problem for image formats like `CL_RGB` where the image element size is not a power of two.

@kpet, I don't think this was intended.  Is the fix to clarify that the computed alignment using the core `CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT` query must be a power of two after converting from pixels to bytes?  The extension text seems to be saying what we want it to say, so I don't think any changes would occur there.

@JohnKesapidesARM, if you need a short-term fix while we work this out could we perhaps perform this check only when the element size is a power of two?  I think the only questionable cases are when the element size is non-power of two.

Reply:
@bashbaug I have updated the test based on the comments. The test is now only doing the checks checking when element_size is a power of two. 
Unfortunately, I had to force push the update to amend the commit message.